File: 2017/24.py

import math, copy, re, hashlib
import itertools as it
# import lib for year 2020

def rl(arr):
	return range(len(arr))

def parse(p):
	return list(map(int, p.split('/')))

def calc_str(ports, used):
	S = 0

	for i in used:
		S += ports[i][0] + ports[i][1]
	return S

def find_ports(first, ports, used):
	found = False
	mx = strm = 0
	for i in rl(ports):
		if i in used:
			continue
		p = ports[i]
		fst = p[0] == first
		snd = p[1] == first
		if fst or snd:
			found = True
			used |= {i}
			curr_s, curr_l = find_ports(p[1] if fst else p[0], ports, used)
			if curr_l > mx:
				mx = curr_l
				strm = curr_s
			elif curr_l == mx and curr_s > strm:
				strm = curr_s
			used -= {i}
	if not found:
		S = calc_str(ports, used)
		return S, len(used)
	return strm, mx


def part_1(data):

	ports = [parse(p) for p in data]
	first = list(filter(lambda x: x[0] == 0, range(ports)))
	ports = list(filter(lambda x: x[0] != 0, ports))
	mx = 0
	for fp in first:
		used = set()
		res, _ = find_ports(fp[1], ports, used)
		res += fp[1]
		if res > mx:
			mx = res

	print(mx)
	print('END OF PART1')
	return

def part_2(data):
	ports = [parse(p) for p in data]
	first = list(filter(lambda i: ports[i][0] == 0, range(len(ports))))
	mx = strm = 0
	for fp in first:
		used = set([fp])
		res, l = find_ports(ports[fp][1], ports, used)
		if l > mx:
			mx = l
			strm = res
		elif l == mx and res > strm:
			strm = res

	print(strm, mx)
	print('END OF PART2')
	return 


if __name__ == '__main__':
	with open('24_input') as f:
		data = f.read()
		data = data.split('\n')
		# data = list(map(int, data.split()))


	part_1(copy.deepcopy(data))
	part_2(copy.deepcopy(data))
	

File: 2017/24.js

const fs = require('fs');
const { exec } = require("child_process");
const { ALPHA_L, ALPHA_U, D4, D6, D8, MOD } = require("../lib");
const { joseph, findCycle } = require("../lib");
const { cache } = require("../lib");
const { bin, float, hex, int, num, oct } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { copy,	entries, in_, inBB, keys, sort, values } = require("../lib");
const { Counter, DD, empty, FastQueue, PriorityQueue, Queue, set, Stack } = require("../lib");
const { areaInt, circumference, manDist, shoelace } = require("../lib");
const { ord, chr, count, debug, disp, disp3, crt, gcd, lcm, modPow, modPowBig, modInv, mod, prod, prodBig, randint, sum, sumBig, transpose } = require("../lib");
const { digits, ints,	floats,	singles,	words,	lines,	table,	groups,	getGroups,	groupsWith,	parse,	parseLine, scanf } = require ('../lib');
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { isSuperset, or, and, xor, sub } = set;
const { getExecStr } = require("../lib/post");
const { combinations, combinations_with_replacement, next_permutation, product } = require("../lib");

const maxBridge = (data, used, curr) => {
	let mx = 0;
	iter(data, ([f, s], i) => {
		if (used[i]) return;
		if (f === curr) {
			used[i] = true;
			mx = max(mx, f + s + maxBridge(data, used, s));
			used[i] = false;
		}
		if (s === curr) {
			used[i] = true;
			mx = max(mx, f + s + maxBridge(data, used, f));
			used[i] = false;
		}
	})
	return mx;
}

function part1(data) {

	let res = 0;
	data = scanf(data, "%d/%d");
	const used = new Array(data.length).fill(false);

	res = maxBridge(data, used, 0);

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART1");
	return;
}

const maxBridgeLen = (data, used, curr, length, sm, mx) => {
	iter(data, ([f, s], i) => {
		if (used[i]) return;
		if (f === curr) {
			used[i] = true;
			maxBridgeLen(data, used, s, length + 1, sm + f + s, mx);
			used[i] = false;
		}
		if (s === curr) {
			used[i] = true;
			maxBridgeLen(data, used, f, length + 1, sm + f + s, mx);
			used[i] = false;
		}
	})
	mx[length] = max(mx[length], sm);
}

function part2(data) {

	let res = 0;
	data = scanf(data, "%d/%d");
	const used = new Array(data.length).fill(false);
	const mx = new DD();
	maxBridgeLen(data, used, 0, 0, 0, mx);
	const mxk = max(...keys(mx))
	res = mx[mxk];
	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART2");
	return;
}

function main() {
	let data = fs.readFileSync("24_input").toString("utf-8");

	part1(data);
	part2(data);
	process.exit(0);
}

main();
